src/v/v8_engine/executor.cc

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "v8_engine/executor.h"
+
+#include "seastarx.h"
+#include "utils/mutex.h"
+#include "vassert.h"
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/smp.hh>
+
+#include <exception>
+
+namespace v8_engine {
+
+namespace internal {
+
+// spsc_queue
+
+spsc_queue::spsc_queue(size_t queue_size)
+  : _items(queue_size)
+  , _ss_mutex_shard(ss::this_shard_id()) {}
+
+void spsc_queue::close() {
+    is_stopped = true;
+    _push_mutex.broken(ss::gate_closed_exception());
+    _has_element_cv.notify_all();
+}
+
+ss::future<> spsc_queue::push(work_item* item) {
+    auto lock = co_await _push_mutex.get_units();
+
+    while (!_items.push(item)) {
+        co_await _is_not_full.wait();
+    }
```

Comment:
After stop executor we need to wait finish all task from queue. So it will be finish after close. Maybe better throw exception if we close queue and do not push element to queue

Reply:
I have test for that